allow vcat on single ChainedVector

stuartthomas25 opened the pull request:
Formerly, `vcat(a::ChainedVector)` failed with an error. This PR makes one-input `vcat` return an unaliased `ChainedVector`, matching the typical Julia behavior that `vcat(a) == a` but `vcat(a) !== a`.

It also adds wrapper-preserving unaliasing for nested `ChainedVector` and `SentinelArray` storage, with regression tests for identity, mutation isolation, missing and sentinel values, heterogeneous chunks, nesting, and inference.

Co-authored by Codex

stuartthomas25 committed:
allow vcat on single ChainedVector

Formerly, `vcat(a::ChainedVector)` would fail with an error.

JoaoAparicio commented:
One observation: The typical Julia behavior is that vcat copies its inputs:

```julia
x = [1,2,3]
vcat(x) === x  # false
```

whereas here
```julia
x = ChainedVector([[1,2,3],[4,5]])
vcat(x) === x  # true
```

ararslan commented on src/chainedvector.jl:
```diff
@@ -629,6 +629,8 @@ Base.@propagate_inbounds function Base.insert!(A::ChainedVector{T, AT}, i::Integ
     return A
 end
 
+Base.vcat(A::ChainedVector{T, AT}) where {T, AT <: AbstractVector{T}} = A
+
 function Base.vcat(A::ChainedVector{T, AT}, arrays::ChainedVector{T, AT}...) where {T, AT <: AbstractVector{T}}
```
I'd suggest folding this into the same method, either with a fast path like
```suggestion
function Base.vcat(A::ChainedVector{T, AT}, arrays::ChainedVector{T, AT}...) where {T, AT <: AbstractVector{T}}
    isempty(arrays) && return Base.unaliascopy(A)
```
or by fixing the source of the error directly by passing `init=0` to `sum` a few lines down. (The latter would require bumping the minimum Julia compat to at least 1.6, but IMO that's fine.)

quinnj committed:
fix(chained-vector): unalias single vcat

Return a wrapper-preserving copy for one-argument vcat.

Teach SentinelArray and nested ChainedVector storage how to make
unaliasing copies. Add coverage for identity, mutation isolation,
missing and sentinel values, heterogeneous chunks, nesting, and
inference.
